Clytwynec/initial work

clytwynec opened the pull request:
What's this PR/Repo for?
---------------------------------
This is the initial work on this repository.  The purpose of this codebase is to write and test [edX specific accessibility rules](https://openedx.atlassian.net/wiki/display/A11Y/edX+Specific+Accessibility+Tests).  These rules are written to work with aXe core, for which ruleset customization is enabled in Bok Choy.

How will this be used?
------------------------------
*  We will be able to use these custom rules in any of our repositories that have bok choy accessibility tests. (Currently, that is edx-platform, edx-analytics-dashboard, and ora-2.)
*  We can version this. 
*  See the README in this PR for some more details about installation, development, and testing.


@jzoldak @benpatterson  Please review. In particular, architecture and general implementation.  

@clrux  This is JavaScript, so I'm hoping that once this lands, the accessibility team will be able to handle the development of new rules internally. Would you mind taking a look at this from that perspective?  The last three commits and the README would be a good place to focus.  

@cptvitamin  FYI

clytwynec commented:
@benpatterson @jzoldak  Is this review in your current sprint?

benpatterson commented:
@clytwynec in the current sprint! I'll take a look now...

benpatterson commented on .gitignore:
```diff
@@ -0,0 +1,27 @@
+coverage
+
+# Grunt intermediate storage (http://gruntjs.com/creating-plugins#storing-task-files)
+.grunt
```
Still reading it, but I didn't notice any grunt-specific stuff in this PR. Is the expectation that this would be absorbed in a higher kind of grunt workflow?

We should be using gulp at edX, since we are already using that on a couple of repositories. But, seems like this is more about the ignoring, so I was curious.

clytwynec replied:
This was the default .gitignore file GitHub generated for node.  I don't have any plans for using either grunt or gulp.  I can remove it from here if it's bothersome.

benpatterson replied:
Ok...no...I was just trying to infer your intention. If it was generated, then that makes sense to me.

benpatterson commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+language: node_js
+node_js:
+    - '0.10'
```
Any reason why not 0.12 as well?

clytwynec replied:
[aXe Core is only testing 0.10](https://github.com/dequelabs/axe-core/blob/master/.travis.yml).   I could test both and see how it goes.

clytwynec replied:
@benpatterson I just updated to address your comments.  Travis builds now run with both 0.10 and 0.12.

benpatterson commented on Makefile:
```diff
@@ -0,0 +1,15 @@
+	$(NODE_BIN)/karma start
+
+quality: requirements
+	$(NODE_BIN)/jshint lib/ test/
```
By the way, we have .jshintrc files in some repos, like [here](https://github.com/edx/edx-analytics-dashboard/blob/master/.jshintrc). I can't remember, but this might be a standard assumption and therefore worthwhile including here. Could you check with Andy and Dennis?

benpatterson replied:
oh, and by saying "including here" I meant "in this repo."

clytwynec replied:
@andy-armstrong @dsjen  Is there a standard .jshintrc file that should be used in edX repos?

benpatterson replied:
Also, does this make command fail (i.e., exit > 0) if violations are found?

clytwynec replied:
Yes to exiting > 0 when there are jshint violations.

benpatterson replied:
Solid. Thanks.

andy-armstrong replied:
@clytwynec There isn't a standard .jshintrc file, unfortunately. Maybe the one in edx-ui-toolkit should be the canonical version, and then we could use Bower to install it into other repos:

https://github.com/edx/edx-ui-toolkit/blob/master/.jshintrc

@dsjen, would that be possible and/or is it a good idea? Obviously it only works for repos that consume the UI Toolkit, but the hope is that eventually the toolkit will benefit all front end repos.

dsjen replied:
Good question!  I think that having a standard version would be possible.  My guess is that we'd need for the repos to point at the right `.jshintrc` file after the bower install.

Do we imagine that different repos might adjust the settings?  I'm wondering if putting the hint file in a https://github.com/edx/cookiecutter-django-ida would make sense or something like https://github.com/edx/edx-lint.

clytwynec replied:
I see the UI toolkit is using gulp for tasks like testing and linting.  I was under the impression that a Makefile was the standard for things like that.  Has that changed?

dsjen replied:
A Makefile didn't seem necessary for the ui toolkit as gulp was able to do everything.  If we wrote a makefile, it would essentially wrap all the gulp tasks.  What do you think?

clytwynec replied:
It seems fine to me, but wasn't sure if it meant that I should be doing something differently here.

dsjen replied:
I see.  If you wanted to use gulp, you could add something like `$(NODE_BIN)/gulp lint`, but it seems like you have that covered with the makefile.

benpatterson replied:
I personally would prefer gulp but defer to @andy-armstrong on the make vs gulp approach. Thoughts, Andy?

andy-armstrong replied:
I asked this same question in the HipChat dev room the other day:

https://edx.hipchat.com/embedded/history/room/100299/2015/10/30?p=1&q=bower%20make&t=rid-100299#e28c3ab4-1a92-4911-8300-77092010ffc1

The consensus seems to be that ``make`` is lowest common denominator which works regardless of what your repo does -- ``paver`` is too Python specific, and ``gulp`` is too front end-specific. For a purely front end repo, I think it makes sense to use ``gulp`` exclusively without wrapping it with ``make``, but I can see the argument for using ``make`` just for consistency across all of our repos.

andy-armstrong replied:
@dsjen I thought about edx-lint initially, but that seems very PyLint oriented at the moment. If we want a common file then having it in the UI Toolkit seems best because that does get installed. However, as you point out, it may be that every repo wants its own subtle variations, in which case having it in the cookiecutter repo makes more sense.

Note: I have a T&L PR to add front end capabilities to the cookiecutter repo:

https://openedx.atlassian.net/browse/TNL-3681

clytwynec replied:
Ok.  Since this is not using gulp to begin with, I think I will stick with Make here to hopefully keep it consistent with other repos.  For the jshintrc file,  does it make sense to copy the one from the UI toolkit here, or somehow install it?

andy-armstrong replied:
I would just copy the one from the UI Toolkit.

dsjen replied:
+1

clytwynec replied:
@andy-armstrong  I didn't see your most recent comment before.  Doesn't the cookie-cutter repo assume that it's a django app?  If that's the case, for things like this and the UI toolkit, they wouldn't be using the same base jshintrc file.  Would it make sense to create another edx-lint style repo, but for javascript quality?

clytwynec replied:
For this, I'll copy from the toolkit. I can update later if we get a standard somewhere. Thanks for the help @dsjen @andy-armstrong

andy-armstrong replied:
@clytwynec That's a good point. I don't know whether we can generalize the cookiecutter repo so that it doesn't assume Python/Django, or whether it makes sense to introduce another repo. I don't particularly like having hundreds of small repos, so my preference would be to keep it within one of the existing ones.

benpatterson commented on README.rst:
```diff
@@ -0,0 +1,78 @@
+Testing guide
+=============
+
+To run tests for the custom ruleset in "dev" mode:
```
This is for testing that one's new ruleset works, is that correct? As opposed to actually using the rules?

clytwynec replied:
Yea.

benpatterson replied:
Ok cool. Could you clarify that?

benpatterson commented on README.rst:
```diff
@@ -0,0 +1,78 @@
+    make test
+
+
+Testing checklist
```
This section is great.

benpatterson commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "karma-spec-reporter": "0.0.20",
+    "phantomjs": "^1.9.18"
+  },
+  "private": true
```
What does this do?

clytwynec replied:
From the docs: "Consider also setting "private": true to prevent accidental publication."  ...  I should remove this line now.

benpatterson commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,150 @@
+        }
+      },
+      evaluate: function(node, options) {
+        return axe.utils.isHidden(node);
```
I don't see the 'i' element being selected like how similar elements are selected in above tests...how is that done for this function?

benpatterson commented on package.json:
```diff
@@ -0,0 +1,40 @@
+  "devDependencies": {
+    "axe-core": "~1.1.0",
+    "jasmine-core": "^2.3.4",
+    "jshint": "2.8.0",
```
Why not `^2.8.0`?

benpatterson commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "karma-jasmine-html-reporter": "~0.1",
+    "karma-jasmine-jquery": "0.1.1",
+    "karma-phantomjs-launcher": "^0.2.1",
+    "karma-spec-reporter": "0.0.20",
```
same here...why pin the version?

benpatterson commented:
I'm curious if you looked at Jasmine 2 for this? Might be easier to pay off that debt sooner rather than later...

clytwynec commented:
@benpatterson  This is already using  `"jasmine-core": "^2.3.4"`.

benpatterson commented:
Ah, sorry I must've been thinking of the other libraries.

benpatterson commented:
Overall this looks good...I'd just want some clarity on the jshintrc file.

clytwynec commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,150 @@
+    },
+    {
+      "id": "icon-aria-hidden",
+      "selector": "i.icon, i.fa, span.icon, span.fa",
```
@benpatterson  RE: "I don't see the 'i' element being selected like how similar elements are selected in above tests...how is that done for this function?"

This is the `icon-aria-hidden` "Rule" which uses the `icon-aria-hidden` "Check".  The `selector` attribute of the rule is where you can define which elements to run the checks on. Elements that match are passed as `node` to the checks' `evaluate` method.  In edX, these are any `<i>` or `<span>` elements with the "fa" or "icon" class.

benpatterson replied:
Ah! Thanks for explaining.

dsjen commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,150 @@
+var customRules = {
+  "rules": [
```
FYI, single quotes are preferred over double via our [style guide](https://github.com/edx/edx-platform/wiki/Javascript-standards-for-the-edx-platform), which eventually points to [google's guide](https://google.github.io/styleguide/javascriptguide.xml?showone=Strings#Strings).

clytwynec replied:
Ok. I will update. Thanks!

dsjen replied:
Ok, thanks! :rabbit:

clytwynec commented:
Updated this with .jshintrc file.

benpatterson commented:
:+1: from me.

discussed with @jzoldak and I'm really all you need from testeng land. You can feel free to merge from my perspective. You can also feel free to get additional thumbs from our gurus like Dennis and AndyA.

dsjen commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+      evaluate: function(node, options) {
+        label = axe.commons.aria.label(node) || '';
```
Declaring these with `var` will prevent polluting the global namespace.  Typically, this is done with one var, e.g.:

```
var label = axe.commons.aria.label(node) || '',
      words = ...;
```

dsjen commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+      evaluate: function(node, options) {
+        href = node.getAttribute('href');
```
Same `var` recommendation as above.

dsjen commented on lib/custom_a11y_rules.js:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+      evaluate: function(node, options) {
+        href = node.getAttribute('href') || '';
```
Same `var` recommendation as above for `href` and `linkTarget`.

dsjen commented on test/spec/checks_spec.js:
```diff
@@ -0,0 +1,153 @@
+                },
+                results;
+
+            var findCheck = function(){
```
Same `var` recommendation as above for `findCheck` and `evaluateCheck`.

dsjen commented on test/spec/checks_spec.js:
```diff
@@ -0,0 +1,153 @@
+            };
+
+            var evaluateCheck = function(){
+                check = findCheck(checkId);
```
Same `var` recommendation as above for `check` and `nodes`--functional scoping here would make sense.

dsjen commented on test/spec/integration_spec.js:
```diff
@@ -0,0 +1,30 @@
+
+    it('rules can be configured with axe-core', function(){
+        axe.configure(rules);
+        configuredRules = axe.getRules(['edx-custom']);
```
Same `var` recommendation as above for `configuredRules ` and `configuredRuleIds`.

dsjen commented on test/spec/integration_spec.js:
```diff
@@ -0,0 +1,30 @@
+
+    it('checks can be configured with axe-core', function(){
+        axe.configure(rules);
+        configuredChecks = axe._audit.checks;
```
Same `var` recommendation as above for `configuredRules ` and `configuredRuleIds`.

dsjen commented on test/spec/integration_spec.js:
```diff
@@ -0,0 +1,30 @@
+    });
+
+    it('does not contain unicode chars', function(){
+        rulesStr = JSON.stringify(rules);
```
Same `var` recommendation as above for `rulesStr ` and `rulesStrIsAscii `.

dsjen commented on test/spec/rules_spec.js:
```diff
@@ -0,0 +1,100 @@
+                failCount = context.expectedResult.failCount,
+                results;
+
+            var sum = function(numsList){
```
Same `var` recommendation as above for `sum ` and `verifyResults `.

dsjen commented on test/spec/rules_spec.js:
```diff
@@ -0,0 +1,100 @@
+            };
+        };
+
+    var specCases = [
```
Same `var` recommendation as above.

dsjen commented:
@clytwynec -- I wrapped up my pass, and it looks great for the most part.  The one thing I noticed was the need to declare JS variables.  Otherwise, they're globals.

dsjen commented:
I just noticed that some indentation is just two spaces (e.g. `lib/custom_a11y_rules.js`).  I'm surprised that linting didn't catch that.

clytwynec commented:
@dsjen Looks like indentation checking is disabled in the [.jshintrc file in edx-ui-toolkit](https://github.com/edx/edx-ui-toolkit/blob/master/.jshintrc#L17).  Should I update that here?

dsjen commented:
Oh, that's interesting!  My vote would be to turn it on, but I don't want to delay your PR with too much deliberating (as you've already been dealing with in this PR).

clytwynec commented:
@dsjen I think I've addressed all of your comments.  I didn't update the .jshintrc file because it would require enabling a legacy option, that could be removed unexpectedly in the future.  I did switch the indentation on `lib/custom_a11y_rules.js` to 4 spaces though.

dsjen commented:
:dancer: :rabbit2: :zap:
